sdk/src/main/scala/com/horizen/account/forger/AccountForgeMessageBuilder.scala

```diff
@@ -102,44 +102,55 @@ class AccountForgeMessageBuilder(
-          log.trace(s"Could not apply tx, reason: ${e.getMessage}")
-          return Success(receiptList, txHashList, cumBaseFee, cumForgerTips)
-        case Failure(e: NonceTooLowException) =>
-          //SHOULD NEVER HAPPEN, but in case just skip this tx
-         log.error(s"******** Could not apply tx for NonceTooLowException ******* : ${e.getMessage}")
-        case Failure(e) =>
-          // just skip this tx but remove any changes caused by the rejected tx
-          log.warn(s"Could not forge tx, reason: ${e.getMessage}", e)
-         //  txsCounter and  blockSize should be reset but (SPOILER ALERT!) they will be removed in a future PR
+      val revisionId = stateView.snapshot
+      val initialBlockGas = blockGasPool.getGas
+
+      try {
+        stateView.applyTransaction(tx, txIndex, blockGasPool, blockContext, finalizeChanges = false) match {
+          case Success(consensusDataReceipt) =>
+            val ethTx = tx.asInstanceOf[EthereumTransaction]
+            val txHash = BytesUtils.fromHexString(ethTx.id)
+
+            receiptList += consensusDataReceipt
+            txHashList += txHash
+
+            val txGasUsed = consensusDataReceipt.cumulativeGasUsed.subtract(cumGasUsed)
+            // update cumulative gas used so far
+            cumGasUsed = consensusDataReceipt.cumulativeGasUsed
+
+            val baseFeePerGas = blockContext.baseFee
+            val (txBaseFeePerGas, txForgerTipPerGas) = GasUtil.getTxFeesPerGas(ethTx, baseFeePerGas)
+            cumBaseFee = cumBaseFee.add(txBaseFeePerGas.multiply(txGasUsed))
+            cumForgerTips = cumForgerTips.add(txForgerTipPerGas.multiply(txGasUsed))
+
+          case Failure(e: GasLimitReached) =>
+            // block gas limit reached
+            // keep trying to fit transactions into the block: this TX did not fit, but another one might
+            log.trace(s"Could not apply tx, reason: ${e.getMessage}")
+            // skip all txs from the same account
+            accountsToSkip += tx.getFrom
+          case Failure(e: FeeCapTooLowException) =>
+            // stop forging because all the remaining txs cannot be executed for the nonce, if they are from the same account, or,
+            // if they are from other accounts, they will have a lower fee cap
+            log.trace(s"Could not apply tx, reason: ${e.getMessage}")
+            return Success(receiptList, txHashList, cumBaseFee, cumForgerTips)
+          case Failure(e: NonceTooLowException) =>
+            //SHOULD NEVER HAPPEN, but in case just skip this tx
+            log.error(s"******** Could not apply tx for NonceTooLowException ******* : ${e.getMessage}")
+          case Failure(e) =>
+            // skip all txs from the same account but remove any changes caused by the rejected tx
+            log.warn(s"Could not forge tx, reason: ${e.getMessage}", e)
+            accountsToSkip += tx.getFrom
+            stateView.revertToSnapshot(revisionId)
```

Comment:
This revert is the only reason we need the `finalizeChanges` switch on `applyTransaction`, correct?
Is there a reason we can't make `applyTransaction` revert everything on its own if this occurs?

Reply:
True. The first implementation actually was as you suggest but after a comment from Sasha, it was changed as it is now.
The rational for this change is to let the forger decide when to call finalizeChanges.

Reply:
Ok, good.